torch/fx/experimental/symbolic_shapes.py

```diff
@@ -274,17 +277,27 @@ def _get_key(self):
         """
         return (len(self.replacements), len(self.divisible))
 
+    # NB: This is only called for input symbolic sizes; intermediate symbolic
+    # sizes are allocated via a different mechanism
     def create_symint(self, name, val):
+        assert val >= 0
         if not HAS_SYMPY:
             raise RuntimeError("Need sympy installed to create symbolic shapes")
 
-        # Currently we don't put 0/1 specialization in guards but perhaps we should
+        # TODO: Put 0/1 specialization in guards
         if val == 0 or val == 1:
             return val
+        # This implements duck-shaping: input sizes that match are assigned
```

Comment:
This is a tradeoff to speed up tracing as we have less symint to handle at the cost of some potential recompilation?

Reply:
Yep

Reply:
It's also not just tracing speed; it's easier to generate kernels with more constraints like this